### PR TITLE
Add search in Backpacks

### PR DESCRIPTION
**Pull request by slprime**
### Add search in Backpacks:
0. Forestry was added earlier by another developer
1. AdventureBackpack2 (also looking for fluids)
2. ProjectRed
3. Minecraft-Backpack-Mod (no highlight backpack because client don't have information about inventory)

image

Others changes:
1. created `IStackFilterProvider` to make it easier to add new mods in the future
2. flag `ItemHighlightingEmptyItemStacks` works with fluid to
3. some of the settings are taken from the client (ItemHighlightingEmptyItemStacks, SearchInGregTechPipes, SearchInEnderIOConduits)

### CODE REVIEW

File: src/main/java/com/gtnh/findit/IStackFilter.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import net.minecraft.inventory.IInventory;
+import net.minecraft.item.ItemStack;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraftforge.fluids.FluidStack;
+
+import com.gtnh.findit.service.itemfinder.FindItemRequest;
+
+public interface IStackFilter {
+
+    public boolean matches(FindItemRequest request);
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnh/findit/IStackFilter.java]: Instance members of interface are always public and don't require access modifiers. Same for methods of `IStackFilterProvider`.

File: src/main/java/com/gtnh/findit/IStackFilter.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.inventory.IInventory;
+import net.minecraft.item.ItemStack;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraftforge.fluids.FluidStack;
+
+import com.gtnh.findit.service.itemfinder.FindItemRequest;
+
+public interface IStackFilter {
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnh/findit/IStackFilter.java:15]: You can mark this interface with `@FunctionalInterface` annotation if you want to emphasize that lambda implementations are supported (Like one for Draconic Evolution).

File: src/main/java/com/gtnh/findit/service/itemfinder/FindItemRequest.java [hunk 3]
```diff
@@ -1,106 +1,102 @@
...
+    private boolean highlightingEmptyItemStacks;
+    private boolean searchInGTPipes;
+    private boolean searchConduits;
 
     public FindItemRequest(ItemStack targetStack) {
         this.targetStack = targetStack;
         this.targetFluidStack = StackInfo.getFluid(targetStack);
+        this.highlightingEmptyItemStacks = FindItConfig.ITEM_HIGHLIGHTING_EMPTY_ITEMSTACKS;
+        this.searchInGTPipes = FindItConfig.SEARCH_IN_GT_PIPES;
+        this.searchConduits = FindItConfig.SEARCH_IN_ENDERIO_CONDUITS;
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnh/findit/service/itemfinder/FindItemRequest.java:37]: Storing and syncing of all these values are performed for respecting client config, am I correctly understanding your logic?

**Review by OneEyeMaker — CHANGES_REQUESTED** [review]

File: src/main/java/com/gtnh/findit/service/itemfinder/FindItemRequest.java (hunk 3 above)
Comment (slprime) [reply src/main/java/com/gtnh/findit/service/itemfinder/FindItemRequest.java:37]: yes. these parameters are more specific to the player than to the server

**Review by OneEyeMaker — APPROVED** [review]

**Pull request merged**